Subcell tags for the mesh, coordinates, and TCI status

nilsdeppe opened the pull request:
## Proposed changes

Exactly what the description says. I will be adding a compute tag for the TCI status in a followup PR so that it gets set automatically depending on the mesh.

### Upgrade instructions

<!--
If this PR makes changes that other people should be aware of when upgrading
their code, describe what they should do between the two UPGRADE INSTRUCTIONS
lines below.
-->
<!-- UPGRADE INSTRUCTIONS -->

<!-- UPGRADE INSTRUCTIONS -->

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `major new feature` if appropriate.

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

nilsdeppe committed:
Add Mesh tag for subcells

nilsdeppe committed:
Add Coordinates and LogicalCoordinates tags for subcell

nilsdeppe committed:
Add TciStatus tag for DG-subcell

kidder commented on tests/Unit/Evolution/DgSubcell/Test_Tags.cpp:
```diff
@@ -29,6 +34,25 @@ template <size_t Dim>
+  TestHelpers::db::test_simple_tag<
+      evolution::dg::subcell::Tags::Coordinates<Dim, Frame::Inertial>>(
+      "InertialCoordinates");
+
```
please run `TestHelpers::db::test_compute_tag` on the compute tag

kidder commented on src/Evolution/DgSubcell/Tags/TciStatus.hpp:
```diff
@@ -0,0 +1,19 @@
+namespace evolution::dg::subcell::Tags {
+/// Stores the status of the troubled cell indicator in the element as a
+/// `Scalar<DataVector>` so it can be observed.
+struct TciStatus : db::SimpleTag {
```
is this going to be a value per grid point or one value for the entire grid?

nilsdeppe replied:
Technically it's one value per element, but I don't know how to visualize that in ParaView, so I've opted for making it a `Scalar<DataVector>` so it's easy to visualize. I expect that once things are tested and working we can avoid adding it to the DataBox by default. I think it's also possible to deduce this info just from the mesh, but again, the path of least resistance (by far) is just writing out a `Scalar<DataVector>`. Does that make sense?

kidder replied:
Yes that makes sense.  I guess my next questions are: 
- Are you printing a size 1 DataMesh or a constant on the active mesh of the element?
- What determines the value printed? (i.e. is it a discrete 0 or 1 depending upon whether or not the Element is troubled, or some continuous value that conveys more information)

nilsdeppe replied:
1. it is the size of the active grid. `N+1` on DG and `2N+1` on subcell
2. it is discrete 1 or 0. I don't really know how to make additional information available easily since each executable has its own set of TCIs

nilsvu replied:
I think this is a reasonable approach since it just works with the current infrastructure. @nilsdeppe Do you think it's feasible to add support for single-valued attributes per element to the observers at some point in the future? I think we'll have more applications for that, e.g. for AMR error estimators, or to visualize wall-time distribution and node distribution.

nilsdeppe replied:
From the spectre writing perspective it's not a problem, I don't know what ParaView supports, that could be a bit trickier.

nilsvu replied:
Ok, I think if writing such things through the observers works then we'll also be able to pass it along to ParaView. It can do things like color each block in a different color by default, so it should be able to handle this. I looked through https://www.xdmf.org/index.php/XDMF_Model_and_Format but didn't immediately find a solution. I'll open an issue.

nilsdeppe replied:
:+1:

kidder reviewed (changes requested):
you can squash

kidder commented on src/Evolution/DgSubcell/Tags/Mesh.hpp:
```diff
@@ -0,0 +1,21 @@
+/// \endcond
+
+namespace evolution::dg::subcell::Tags {
+/// The mesh on the subcells
```
I couldn't find this documentation; it appears the Tags namespace in subcell is not documented?

kidder replied:
okay, I found the documentation for the Mesh tag; but the Tags namespace does not appear to be documented

kidder commented on src/Evolution/DgSubcell/Tags/TciStatus.hpp:
```diff
@@ -0,0 +1,19 @@
+
+namespace evolution::dg::subcell::Tags {
+/// Stores the status of the troubled cell indicator in the element as a
+/// `Scalar<DataVector>` so it can be observed.
```
hmm this one shows up, so I don't know why the other one doesn't

nilsvu commented on src/Evolution/DgSubcell/Tags/Mesh.hpp:
```diff
@@ -0,0 +1,21 @@
+/// The mesh on the subcells
+template <size_t VolumeDim>
+struct Mesh : db::SimpleTag {
+  using type = ::Mesh<VolumeDim>;
```
I think it would be good to add a `name` function for this and the other subcell tags so the names include the word `subcell`. For example, I think this tag could have the name `SubcellMesh`. That would help disambiguate these tags from the non-subcell items in the box.

nilsdeppe replied:
I don't really see a need for that with these tags. I don't think they're ever retrieved by name, and if the active grid is FD/FV then it's not clear what prefix (if any) should be used.

nilsvu replied:
This is no hill I want to die on, though when printing DataBox contents and it just says "Mesh" twice I would find that confusing. Since this is in a `dg::subcell` namespace I feel like `SubcellMesh` would describe this well...

nilsvu commented on src/Evolution/DgSubcell/Tags/Coordinates.hpp:
```diff
@@ -0,0 +1,45 @@
+
+/// The logical coordinates on the subcell grid
+template <size_t VolumeDim>
+struct LogicalCoordinates : Coordinates<VolumeDim, Frame::Logical>,
```
`LogicalCoordinatesCompute`?

nilsdeppe replied:
This is the same name that `domain/LogicalCoordinates` uses :shrug:

nilsvu replied:
Per our recent conventions that one should probably also changed then (not in this PR of course)

fmahebert reviewed (commented):
I have no additional requests

nilsdeppe committed:
Document subcell Tags and OptionTags namespaces

nilsdeppe commented on src/Evolution/DgSubcell/Tags/Coordinates.hpp:
```diff
@@ -0,0 +1,45 @@
+template <size_t Dim, typename Frame>
+struct Coordinates : db::SimpleTag {
+  static std::string name() noexcept {
+    return get_output(Frame{}) + "Coordinates";
```
Prefixing this with `Subcell` would break visualizations, and I can't really spend time overhauling all the visualization scripts to support subcell at that level

nilsvu replied:
I'm fine leaving this as-is for now to avoid refactoring anything. Though I think it would be quite straight-forward: Since you have two different tags for the subcell/dg coordinates anyway you have to switch between them in your observation event, at which point you can just pick a consistent label for the dataset in the output file.

Question: Will the subcell code actually write into the same volume data as the DG code?

nilsdeppe replied:
No, not nearly that straightforward. I have a wrapper event that calls into the impl of the generic fields event. It is that generic event that decides the name. I'd have to do some refactoring of the generic event, then the subcell event, and then I'd have to refactor `Render1D`, and also `GenerateXdmf`. It's a pretty big project.

Yes, DG and subcell data are written into the same subfile. Writing into different subfiles will require an even bigger rewrite of the visualization scripts. I currently see no benefit to writing into different subfiles and lots of drawbacks.

nilsdeppe commented:
I've made the following changes (I squashed because they're rather tiny):
- add `name() { return "Subcell(Mesh); }` function (adjusted test accordingly)
- added the `test_compute_tag<LogicalCoordinatesCompute>`
- rename `LogicalCoordinates` to `LogicalCoordinatesCompute`

Thanks for the reviews :)

kidder commented on src/Evolution/DgSubcell/Tags/Tags.hpp:
```diff
@@ -0,0 +1,19 @@
+
+namespace fd {
+/// \ingroup DgSubcellGroup
+/// \brief %Tags for the DG-subcell finite difference solversolver
```
oops

nilsdeppe replied:
Fixed :)

kidder reviewed (changes requested):
you can squash
